src/lib/mqtt/MqttConnectionManager.cpp

```diff
@@ -0,0 +1,357 @@
+      semGive();
+      return;
+    }
+
+    mosquitto_publish_callback_set(c.mosq, mqttOnPublishCallback);
+
+    LM_T(LmtMqttNotif, ("Connecting to MQTT Broker at %s:%d", host.c_str(), port));
+
+    int resultCode = mosquitto_connect(c.mosq, host.c_str(), port, MQTT_DEFAULT_KEEPALIVE);
+    if (resultCode != MOSQ_ERR_SUCCESS)
+    {
+      LM_E(("Runtime Error (could not connect to MQTT Broker %s:%d (%d): %s)", host.c_str(), port, resultCode, mosquitto_strerror(resultCode)));
+      mosquitto_destroy(c.mosq);
+      semGive();
+      return;
+    }
+
+    LM_T(LmtMqttNotif, ("Successfully connected to MQTT Broker"));
+
+    // Starts the client loop in its own thread. The client loop processes the network traffic.
+    // According to documentation (https://mosquitto.org/api/files/mosquitto-h.html#mosquitto_threaded_set):
+    // "When using mosquitto_loop_start, this [mosquitto_threaded_set] is set automatically."
+    mosquitto_loop_start(c.mosq);
+
+    connections[endpoint] = c;
+  }
+  else
+  {
+    // Exist: increase the counter
+    connections[endpoint].counter++;
+
+    LM_T(LmtMqttNotif, ("Existing MQTT broker connection for %s:%d (counter is now: %d)", host.c_str(), port, connections[endpoint].counter));
+  }
+
+  semGive();
+}
+
+
+
+/* ****************************************************************************
+*
+* MqttConnectionManager::connect -
+*
+*/
+void MqttConnectionManager::connect(const std::string& url)
+{
+  std::string  host;
+  int          port;
+  std::string  path;
+  std::string  protocol;
+
+  if (!parseUrl(url, host, port, path, protocol))
+  {
+    LM_E(("Runtime Error (malformed URL: '%s')", url.c_str()));
+  }
+  else
+  {
+    connect(host, port);
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* MqttConnectionManager::disconnect -
+*/
+void MqttConnectionManager::disconnect(const std::string& host, int port)
+{
+  std::string endpoint = getEndpoint(host, port);
+
+  semTake();
+  if (connections.find(endpoint) == connections.end())
+  {
+    // Doesn't exist: this is an error situation
+    LM_E(("Runtime Error (MQTT broker connection for %s:%d doesn't exist)", host.c_str(), port));
+  }
+  else
+  {
+    connections[endpoint].counter--;
+
+    MqttConnection c = connections[endpoint];
+
+    if (c.counter == 0)
+    {
+      // Not used anymore: destroy it
+      LM_T(LmtMqttNotif, ("Disconnecting from MQTT Broker at %s:%d", host.c_str(), port));
+      int resultCode = mosquitto_disconnect(c.mosq);
+      if (resultCode != MOSQ_ERR_SUCCESS)
+      {
+        LM_E(("Runtime Error (could not disconnect from MQTT Broker (%d): %s)", resultCode, mosquitto_strerror(resultCode)));
+      }
+      else
+      {
+        LM_T(LmtMqttNotif, ("Successfully disconnected from MQTT Broker"));
+      }
+      mosquitto_loop_stop(c.mosq, false);
+      mosquitto_destroy(c.mosq);
```

Comment:
Here we also have to add the effective removal from the hashmap of the MqttConnection object.

Reply:
Done in the second pass at L318-L321

NTC